source/extensions/transport_sockets/alts/alts_tsi_handshaker.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "source/extensions/transport_sockets/alts/alts_proxy.h"
+#include "source/extensions/transport_sockets/alts/tsi_frame_protector.h"
+
+#include "grpcpp/channel.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+constexpr std::size_t kAltsMinFrameSize = 16 * 1024;
+
+struct AltsHandshakeResult {
+  TsiFrameProtectorPtr frame_protector;
+  std::string peer_identity;
+  std::string unused_bytes;
+};
+
+// Manages a single ALTS handshake.
+class AltsTsiHandshaker {
+public:
+  using OnNextDone = std::function<void(
+      absl::Status status, void* handshaker, const unsigned char* bytes_to_send,
+      size_t bytes_to_send_size, std::unique_ptr<AltsHandshakeResult> handshake_result)>;
+
+  static std::unique_ptr<AltsTsiHandshaker>
+  createForClient(std::shared_ptr<grpc::Channel> handshaker_service_channel);
+  static std::unique_ptr<AltsTsiHandshaker>
+  createForServer(std::shared_ptr<grpc::Channel> handshaker_service_channel);
+
+  // Get the next ALTS handshake message for the peer.
+  absl::Status next(void* handshaker, const unsigned char* received_bytes,
+                    size_t received_bytes_size, OnNextDone on_next_done);
+
+  // Exposed for testing purposes only.
+  absl::StatusOr<std::unique_ptr<AltsHandshakeResult>>
+  getHandshakeResult(const grpc::gcp::HandshakerResult& result, absl::string_view received_bytes,
```

Comment:
That makes sense then to not call it for "ForTest". But  can you point to where we call it explicitly from a test e.g. such that it helps with coverage (I might have missed it). Otherwise if we just use it via the public interface in test, making it private sgtm.